orttraining/orttraining/core/agent/training_agent.cc

```diff
@@ -2,135 +2,80 @@
-    if (onnxruntime::contrib::OrtTasks::GetInstance().ForwardOutputsIsValid()) {
-      ORT_ENFORCE(!status.IsOK());
-      // signal main thread for background thread completion
-      onnxruntime::contrib::OrtTasks::GetInstance().SetForwardOutputs(status, {});
+    ORT_ENFORCE(utils::InitializeFeedFetchCopyInfo(session_state, *fw_feeds_fetches_manager_) == Status::OK());
+  }
+  {
+    FeedsFetchesManager::Create(bw_feed_names, bw_fetches_names, session_state.GetOrtValueNameIdxMap(), bw_feeds_fetches_manager_);
+    auto& fetch_info = bw_feeds_fetches_manager_->GetMutableFetchesDeviceCopyInfo();
+    for (size_t i = 0, end = bw_fetches_names.size(); i < end; ++i) {
+      fetch_info[i].target_device = bw_outputs_device_info[i];
     }
-  },
-                               std::move(setup_future), std::cref(run_options), std::ref(io_binding));
 
-  run_id = std::hash<std::thread::id>()(bg_thread.get_id());
-  {
-    std::lock_guard<std::mutex> lock(bg_threads_mutex_);
-    bg_threads_[run_id] = std::move(bg_thread);
+    ORT_ENFORCE(utils::InitializeFeedFetchCopyInfo(session_state, *bw_feeds_fetches_manager_) == Status::OK());
   }
 
-  onnxruntime::contrib::OrtTasks::GetInstance().CreateBackgroundTask(run_id);
-
-  LOGS(*inference_session_.GetLogger(), VERBOSE) << "InferenceSession::Forward() call created a task with run_id " << run_id;
-
-  // background task is setup, unblock background thread to continue
-  setup_promise.set_value();
-
-  // Wait for data/signal from
-  // 1. Yield op, if the bg thread sucessfully reached Yield's signal point
-  // 2. The end of bg thread, if it hit execptions and returned earlier
-  auto forward_outputs = onnxruntime::contrib::OrtTasks::GetInstance().WaitForForwardOutputs(run_id);
-  const Status& forward_status = forward_outputs.first;
-  user_outputs = std::move(forward_outputs.second);
-
-  // background thread has completed without hitting Yield Op
-  if (!forward_status.IsOK()) {
-    std::thread thread;
-    {
-      std::lock_guard<std::mutex> lock(bg_threads_mutex_);
-      std::swap(thread, bg_threads_[run_id]);
-      bg_threads_.erase(run_id);
+  size_t end_point;
+  size_t break_point = 0;
+  const SequentialExecutionPlan& seq_exec_plan = *(session_state.GetExecutionPlan());
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  end_point = exec_plan_vec.size() - 1;
+  for (size_t program_counter = 0; program_counter < exec_plan_vec.size(); program_counter += 1) {
+    const auto& node_exec_plan = exec_plan_vec[program_counter];
+    auto node_index = node_exec_plan.node_index;
+    if (session_state.GetKernel(node_index)->KernelDef().OpName() == "YieldOp") {
+      break;
     }
-    ORT_ENFORCE(thread.joinable());
-    thread.join();
-    onnxruntime::contrib::OrtTasks::GetInstance().RemoveTask(run_id);
-    return forward_status;
+    break_point += 1;
   }
 
-  return Status::OK();
+  fw_program_counter_end_ = break_point - 1;
```

Comment:
If we make the counter_end_ exclusive, we can skip the "-1" in the index counting. 


end_point = exec_plan_vec.size() - 1;
fw_program_counter_end_ = break_point - 1;